High-Scores: Add immutability test

jeffdparker opened the pull request:
I don't know the syntax, but would like to add a test that asks for top three (causing a sort) followed by a latest to see if the program sorted in place.

rpottsoh commented on exercises/high-scores/canonical-data.json:
```diff
@@ -77,6 +77,14 @@
           },
           "expected": [40]
         }
```
```suggestion
        },
```

rpottsoh commented:
This proposed addition seems similar to what was removed in #1459.

jeffdparker commented:
Sorry to have overlooked this history.  I am back to request a test that calls latest() twice, due to a solution that used pop() to implement it.

rpottsoh commented:
No problem.  Can you post the code that has driven you to want a test added?  It may be that the test need only be added for that track.  But lets see what you have.  I don't quite follow your reasoning for a new case at this point.

jeffdparker commented:
I can't find it in my lists, but briefly the solution stores the results in a list, and then calls pop(), which removes and returns the last item on the list.  

Like the issue with sorting, this method destroys the integrity of the data.

sshine commented:
This does sound like you want a track-specific test.

The current canonical data describes the expected output of running functions in isolation.

So if you'd like to test that a function is immutable, our current canonical data format assumes them to be and doesn't really support this explicitly.

Also, I don't think it should. If mentors/maintainers of other tracks disagree, feel free to object to this.

But I do think you should add a test for this in your track if this is something you expect to occur again.

F3PiX commented:
It looks like a perfect Talking Point to add to the Mentor Notes, rather than adding a  test.

jeffdparker commented:
There is already note about this in the python mentoring guide.  

The problem of making sure the representation provides an accurate version of the state seems like the kind of thing that would be useful to test, especially as we try to automate what we can.

rpottsoh commented:
Perhaps this would be acceptable:
```json
       {
          "description": "Recent after a call to sort",
          "property": "latest",
          "input": {
            "personalTopThree": {
              "scores": [10, 20, 30, 40, 25]
            }
          },
          "expected": [25]
        }
```

rpottsoh commented:
My implementation of this test would look like:
```pascal
procedure THighScoresTest.Recent_after_a_call_to_sort;
var
  Scores: IScores;
begin
  Scores := NewScores([10, 20, 30, 40, 25]);
  Scores := NewScores(Scores.personalTopThree);
  Assert.AreEqual(25, Scores.latest);
end;
```
Only `latest` is tested.

petertseng commented:
as you have seen, and as I will also now point you to https://github.com/exercism/problem-specifications/issues/1225 to see, the schema optimises for `f(input) = output`; support for sending two (or more) messages to the same receiver was not given first-priority consideration in the design. If you would like to see some examples of how it was done, take a look at https://github.com/exercism/problem-specifications/blob/master/exercises/circular-buffer/canonical-data.json

this comment should not be taken as a recommendation (either positive or negative) on whether this case should be added

petertseng commented:
EBWODP:

> If you would like to see some examples of how it was done, take

and then I only provided one.

I also meant to point out react as an example, and possibly zipper.

ErikSchierboom commented:
> Also, I don't think it should. If mentors/maintainers of other tracks disagree, feel free to object to this.

I agree with @sshine. I understand the thoughts behind the test case, but I'm not in favor of adding it to the canonical data.

sshine commented:
> Is it difficult to make two calls inside a test?

No, but do you mean inside a test file for an exercise on a language track, or in canonical-data.json?

> Is the reluctance to test this driven by technology or by pedagogy?

Technology, entirely!

My main point of objection would be that [the schema of canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/canonical-schema.json) is challenged up to a point where we wrap a bunch of human-readable sentences in JSON and rule out automated unit test generation.

So while some tracks may benefit from the added tests, other tracks for which automated test generation based on canonical-data.json can no longer run their test generator. This exercise was implemented by the following tracks: scala, ecmascript, delphi, nim, javascript, fsharp, csharp, python, ruby. According to exercism/discussions#155, scala, csharp and ruby have test generators.

I think your (property) test should be added to canonical-data.json.

But I'd prefer if canonical-data.json supported properties, then.

I suppose we could

1. Add a property test to high-scores' canonical-data.json that may break test generators, work on making canonical-data.json compatible with properties (one way or another, as continued in #1225), and then refactor this property test.
   Pros: All tracks benefit from the extra test case. Test is stored in central location.
   Cons: May break test generators. Format of canonical-data.json is abused.
2. Add track-specific tests to high-scores, work on making canonical-data.json compatible with properties (one way or another, as continued in #1225), add the property test to high-scores' canonical-data.json and sync the test version of the track(s) that added the local test case.
   Pros: Doesn't break test generators. Keep format of canonical-data.json pure.
   Cons: Track-specific tests don't propagate to other tracks. We may never resolve the discussion in #1225.
3. Something else entirely...

As a last note, I have to admit that I already abused canonical-data.json to store property-based tests in [dnd-character](https://github.com/exercism/problem-specifications/blob/master/exercises/dnd-character/canonical-data.json); I reasoned that because this was a new exercise, it wouldn't break existing test generators, and because it operates on randomness, testing becomes very track-specific (mocking? injecting seed? property-based?), making test generation difficult if not impossible. But this still violates at least one argument I'm using against adding a property-based test to high-scores.

I'm not sure how we proceed from here. I withdraw my objection and am open to proceed in any way.

ErikSchierboom commented:
@jeffdparker Are you still interested in working on this?

ErikSchierboom commented:
Great. Let's discuss what needs to happen when you're back.

kotp commented:
> Yes, I am interested. I’m out of town now, but will be back at my desk
Monday

@jeffdparker Which monday? ;)

jeffdparker commented:
I am back, but I thought I saw that the bug was closed.

I assume that what is needed is a new test: something like this

    def test_do_not_reorder(self):

        scores = [10, 30, 90, 30, 100, 20, 10, 0, 30, 40, 40, 70, 33]


        expected = 33

        self.assertEqual(latest(scores), expected)


        expected = [100, 90, 70]

        self.assertEqual(personal_top_three(scores), expected)


        expected = 33

        self.assertEqual(latest(scores), expected)


        expected = 100

        self.assertEqual(personal_best(scores), expected)


        expected = 33

        self.assertEqual(latest(scores), expected)

I don't know how to insert such a test, but would be happy to learn

jdp

On Sun, Jan 23, 2022 at 8:13 PM Victor Goff ***@***.***>
wrote:

> Yes, I am interested. I’m out of town now, but will be back at my desk
> Monday
>
> @jeffdparker <https://github.com/jeffdparker> Which monday? ;)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1019673487>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CNMQ5NHI7EKCPFJ6RTUXS7VHANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

glennj commented:
This is what [groovy does](https://github.com/exercism/groovy/blob/main/exercises/practice/high-scores/src/test/groovy/HighScoresSpec.groovy#L84):
```
    def "Personal top three does not mutate"() {
        given:
        def hs = new HighScores(scores)
        def top3 = hs.personalTopThree()

        expect:
        hs.latest() == expected

        where:
        scores           || expected
        [40, 20, 10, 30] || 30
    }
```
But I don't know how to express that with the canonical-data-schema

ErikSchierboom commented:
@jeffdparker A separate PR added two test cases to verify immutability: https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110 Could you check to see if those work for your suggestion too?

jeffdparker commented:
Erik -

Yes, those are the two test cases I'm concerned about.  I see the language,
but I don't see how that is translated into test cases.

jdp

PS Briefly, there are three calls: latest(), ..._best() and
 ..._top_three().  There is no temptation to modify the list in
implementing latest(), but there is for the other two, and we should check
that calling each does not modify the list, which we can measure by calling
latest()

On Tue, Jan 25, 2022 at 4:42 AM Erik Schierboom ***@***.***>
wrote:

> @jeffdparker <https://github.com/jeffdparker> A separate PR added two
> test cases to verify immutability:
> https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110
> Could you check to see if those work for your suggestion too?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1021101535>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CJULLKN5S5XFMJE5ADUX2EBVANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

glennj commented:
> @jeffdparker A separate PR added two test cases to verify immutability: https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110 Could you check to see if those work for your suggestion too?

Those cases simply hint to the track test implementers that the top-three functions need to be called first? Or does it imply some ordering to the tests?

Does the schema allow something like this (borrowing from Spock DSL):
```json
        {
          "uuid": "2df075f9-fec9-4756-8f40-98c52a11504f",
          "description": "Latest score after personal top scores",
          "given": {
            "comments": [
              "After invoking the `personalTopThree` property,",
              "the latest score should remain the same."
            ],
            "properties": [
              "personalTopThree"
            ]
          },
          "property": "latest",
          "input": {
            "scores": [70, 50, 20, 30]
          },
          "expected": 30,
          "scenarios": [
            "immutable"
          ]
        },
        {
          "uuid": "809c4058-7eb1-4206-b01e-79238b9b71bc",
          "description": "Scores after personal top scores",
          "given": {
            "comments": [
              "After invoking the `personalTopThree` property,",
              "the scores should remain in the same order."
            ],
            "properties": [
              "personalTopThree"
            ]
          },
          "property": "scores",
          "input": {
            "scores": [30, 50, 20, 70]
          },
          "expected": [30, 50, 20, 70],
          "scenarios": [
            "immutable"
          ]
        }
```

jeffdparker commented:
Glen -

Yes, that looks more like it, but these are testing one API two different
ways, rather than testing the two APIs.

Testing latest() is a standin for testing scores.  Not as accurate, but has
the advantage that it is implemented, and is strong enough to detect
sorting, which is the principal problem.

The specifications you show test ...top_three, but do not test ...best(),
which should also be checked.

jdp


On Tue, Jan 25, 2022 at 9:55 AM Glenn Jackman ***@***.***>
wrote:

> @jeffdparker <https://github.com/jeffdparker> A separate PR added two
> test cases to verify immutability:
> https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110
> Could you check to see if those work for your suggestion too?
>
> Those cases simply hint to the track test implementers that the top-three
> functions need to be called first? Or does it imply some ordering to the
> tests?
>
> Does the schema allow something like this (borrowing from Spock DSL):
>
>         {
>           "uuid": "2df075f9-fec9-4756-8f40-98c52a11504f",
>           "description": "Latest score after personal top scores",
>           "given": {
>             "comments": [
>               "After invoking the `personalTopThree` property,",
>               "the latest score should remain the same."
>             ],
>             "properties": [
>               "personalTopThree"
>             ]
>           },
>           "property": "latest",
>           "input": {
>             "scores": [70, 50, 20, 30]
>           },
>           "expected": 30,
>           "scenarios": [
>             "immutable"
>           ]
>         },
>         {
>           "uuid": "809c4058-7eb1-4206-b01e-79238b9b71bc",
>           "description": "Scores after personal top scores",
>           "given": {
>             "comments": [
>               "After invoking the `personalTopThree` property,",
>               "the scores should remain in the same order."
>             ],
>             "properties": [
>               "personalTopThree"
>             ]
>           },
>           "property": "scores",
>           "input": {
>             "scores": [30, 50, 20, 70]
>           },
>           "expected": [30, 50, 20, 70],
>           "scenarios": [
>             "immutable"
>           ]
>         }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1021400064>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CJNJP7OJQW6DUQZYZLUX3IYJANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ErikSchierboom commented:
> Does the schema allow something like this (borrowing from Spock DSL):

It doesn't (and I'd rather not go there :)).

> Those cases simply hint to the track test implementers that the top-three functions need to be called first? Or does it imply some ordering to the tests?

The property name should be read as follows: "latestAfterTopThree" implies that `topThree` is called first, and then `latest` is called. 

> The specifications you show test ...top_three, but do not test ...best(),
which should also be checked.

I think it would be fine to also have a test like "latestAfterPersonalBest". @SleeplessByte IIRC you reviewed https://github.com/exercism/problem-specifications/pull/1767 which added the array mutation cases. Thoughts on adding more immutable cases?

kotp commented:
Just to interject how I mentor this exercise in Ruby, we talk about if a set of highscores should be immutable, how to make them immutable, and then how to update scores if necessary and why.

So I would ask "Why are we testing for immutability?  Why does it need to be immutable?"

jeffdparker commented:
Your most recent score should not change if you check your best score or
your top three

jdp

On Wed, Jan 26, 2022 at 12:47 AM Victor Goff ***@***.***>
wrote:

> Just to interject how I mentor this exercise in Ruby, we talk about if a
> set of highscores should be immutable, how to make them immutable, and then
> how to update scores if necessary and why.
>
> So I would ask "Why are we testing for immutability? Why does it need to
> be immutable?"
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1021944930>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6COP57ZDRKY2JWYIXJLUX6RI3ANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
- Jeff Parker

kotp commented:
> Your most recent score should not change if you check your best score or your
> top three jdp

That is true, if that is the only thing that was done, however it may be argued
that it should change if there was an update done.  So, I am wondering if
leaving that decision alone leaves room for mentoring and different approaches?

Is this causing an over specification?

This is a practice exercise, rather than a concept exercise, and so we should be
careful of that.

(Also, please cleanup/curate e-mail responses.  It is making top post replies when I
get these e-mail responses, which is making it a little bit difficult to follow.
This will provide some kindness to the reader, even if, unfortunately, it might
require a little bit more care for the author. For example, are you responding
to the first question or the second question? (and so it may even return some of
that kindness to the author in better responses as it is better understood
conversationally what exactly was responded to in that way.)  Currently, I have
to guess as the reader, because the response is out of order.)  (Sorry for the rant... )

jeffdparker commented:
jdp



        Your most recent score should not change if you check your

        best score or your top three


    Victor


    That is true, if that is the only thing that was done, however it

    may be argued that it should change if there was an update done.

    So, I am wondering if leaving that decision alone leaves room for

    mentoring and different approaches?


The problem does not provide any mechanism for updates.  As noted, it

is a simple problem: give me the most recent (last) element of a list,

give me the maximum of the list, and give me the three largest.


A common attempted solution to the problem sorts the list in place

to find the maximum and top three.  Sorting in place can move

the last element, and thus breaks the recent() call.


jdp

ErikSchierboom commented:
> A common attempted solution to the problem sorts the list in place
> to find the maximum and top three.  Sorting in place can move
> the last element, and thus breaks the recent() call.

My proposal would be to add two new test cases. They'll basically be copied of the [existing immutability test cases](https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110), but with "AfterTopThree" changesd to "AfterPersonalBest". @jeffdparker would you be willing to submit a PR?

ErikSchierboom commented:
@jeffdparker Just in case you are unfamiliar with creating PRs, we have a page with some nice links on how to do that: https://exercism.org/docs/building

Once you know how to create a PR, the second step is to create the actual PR. The file you'll need to edit is https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json. What you want to do is add two new test cases (right after https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110):

```json
{
  "uuid": "ddb0efc0-9a86-4f82-bc30-21ae0bdc6418",
  "description": "Latest score after personal best",
  "property": "latestAfterBest",
  "input": {
    "scores": [70, 50, 20, 30]
  },
  "expected": 30,
  "scenarios": [
    "immutable"
  ]
},
{
  "uuid": "6a0fd2d1-4cc4-46b9-a5bb-2fb667ca2364",
  "description": "Scores after personal best",
  "property": "scoresAfterBest",
  "input": {
    "scores": [30, 50, 20, 70]
  },
  "expected": [30, 50, 20, 70],
  "scenarios": [
    "immutable"
  ]
}
```

These are basically copies of the existing immutability tests. The only thing that might be nice to change are the scores used in the test cases, to have them differ from the existing immutability tests. I'll leave that up to you.

N.B. The UUIDs were generated using https://www.uuidgenerator.net/version4.

jeffdparker commented:
Erik -

Thanks - I'm running through the Github course - or perhaps I should say
I'm stumbling through it.  It's a much smoother introduction than the 200
page Atria book.

jdp



On Fri, Jan 28, 2022 at 1:09 AM Erik Schierboom ***@***.***>
wrote:

> @jeffdparker <https://github.com/jeffdparker> Just in case you are
> unfamiliar with creating PRs, we have a page with some nice links on how to
> do that: https://exercism.org/docs/building
>
> Once you know how to create a PR, the second step is to create the actual
> PR. The file you'll need to edit is
> https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json.
> What you want to do is add two new test cases (right after
> https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110
> ):
>
> {
>   "uuid": "ddb0efc0-9a86-4f82-bc30-21ae0bdc6418",
>   "description": "Latest score after personal best",
>   "property": "latestAfterBest",
>   "input": {
>     "scores": [70, 50, 20, 30]
>   },
>   "expected": 30,
>   "scenarios": [
>     "immutable"
>   ]
> },
> {
>   "uuid": "6a0fd2d1-4cc4-46b9-a5bb-2fb667ca2364",
>   "description": "Scores after personal best",
>   "property": "scoresAfterBest",
>   "input": {
>     "scores": [30, 50, 20, 70]
>   },
>   "expected": [30, 50, 20, 70],
>   "scenarios": [
>     "immutable"
>   ]
> }
>
> These are basically copies of the existing immutability tests. The only
> thing that might be nice to change are the scores used in the test cases,
> to have them differ from the existing immutability tests. I'll leave that
> up to you.
>
> N.B. The UUIDs were generated using https://www.uuidgenerator.net/version4
> .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1023977988>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CIXV4D2PVWFKCTHTMDUYJFKRANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jeffdparker commented:
Erik -

So I have a question.  I need to fork, edit the file, and then post a Pull
Request, yes?

I have already forked the area some time ago, and I'm lots of edits
behind.  I assume at this point I need to delete the repository, yes?

jdp

On Fri, Jan 28, 2022 at 1:09 AM Erik Schierboom ***@***.***>
wrote:

> @jeffdparker <https://github.com/jeffdparker> Just in case you are
> unfamiliar with creating PRs, we have a page with some nice links on how to
> do that: https://exercism.org/docs/building
>
> Once you know how to create a PR, the second step is to create the actual
> PR. The file you'll need to edit is
> https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json.
> What you want to do is add two new test cases (right after
> https://github.com/exercism/problem-specifications/blob/main/exercises/high-scores/canonical-data.json#L87-L110
> ):
>
> {
>   "uuid": "ddb0efc0-9a86-4f82-bc30-21ae0bdc6418",
>   "description": "Latest score after personal best",
>   "property": "latestAfterBest",
>   "input": {
>     "scores": [70, 50, 20, 30]
>   },
>   "expected": 30,
>   "scenarios": [
>     "immutable"
>   ]
> },
> {
>   "uuid": "6a0fd2d1-4cc4-46b9-a5bb-2fb667ca2364",
>   "description": "Scores after personal best",
>   "property": "scoresAfterBest",
>   "input": {
>     "scores": [30, 50, 20, 70]
>   },
>   "expected": [30, 50, 20, 70],
>   "scenarios": [
>     "immutable"
>   ]
> }
>
> These are basically copies of the existing immutability tests. The only
> thing that might be nice to change are the scores used in the test cases,
> to have them differ from the existing immutability tests. I'll leave that
> up to you.
>
> N.B. The UUIDs were generated using https://www.uuidgenerator.net/version4
> .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1023977988>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CIXV4D2PVWFKCTHTMDUYJFKRANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ynfle commented:
> Erik - So I have a question. I need to fork, edit the file, and then post a Pull Request, yes? I have already forked the area some time ago, and I'm lots of edits behind. I assume at this point I need to delete the repository, yes? jdp

@jeffdparker you don't need to do that. You would just update your fork `git checkout main && git pull && git push`, rebase on the updated main brach `git checkout patch-2 && git rebase main`, make the necessary changes and push `git push`.

Those commands *should* work if you have upstream and your fork as a remote.

jeffdparker commented:
I would prefer to use the GIT CLI, but I had no luck following the
instructions using {brew} or {port}

I found a CLI command to rebase the branch, and used that instead.

I have inserted the text as suggested, and issued a pull request.

The tool tells me there are no conflicts - is there a way for me to test
the changes?  A lint for specifications?

And what is needed to make the required tests appear?

jdp

On Sun, Jan 30, 2022 at 2:46 AM ynfle ***@***.***> wrote:

> Erik - So I have a question. I need to fork, edit the file, and then post
> a Pull Request, yes? I have already forked the area some time ago, and I'm
> lots of edits behind. I assume at this point I need to delete the
> repository, yes? jdp
>
> @jeffdparker <https://github.com/jeffdparker> you don't need to do that.
> You would just update your fork git checkout main && git pull && git push,
> rebase on the updated main brach git checkout patch-2 && git rebase main,
> make the necessary changes and push git push.
>
> Those commands *should* work if you have upstream and your fork as a
> remote.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1025106697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6COZ4KI5EKUUF6UKDQTUYUCIRANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ynfle commented:
If you push, it should automatically update the PR

jeffdparker commented:
Well, OK - I've hit

Pull request successfully merged and closed



On Sun, Jan 30, 2022 at 9:00 PM ynfle ***@***.***> wrote:

> If you push, it should automatically update the PR
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1025354195>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CLOJ2DZX4R7G3SHAU3UYYCNPANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ErikSchierboom commented:
I think ynfle's suggestion has one tiny bug, and that is that `git push` will fail after rebasing. Once you've rebased, you'll have to do `git push --force`. The fact that the contents in this PR are not updated makes sense as the push will have failed to update it. Could you try force pushing (note: this is generally discouraged, but necessary are rebasing a branch and pushing to an existing remote branch).

jeffdparker commented:
Erik

We are out of town: I will take a look at this later this weekend

Jeff

On Tue, Feb 1, 2022 at 1:25 AM Erik Schierboom ***@***.***>
wrote:

> I think ynfle's suggestion has one tiny bug, and that is that git push
> will fail after rebasing. Once you've rebased, you'll have to do git push
> --force. The fact that the contents in this PR are not updated makes
> sense as the push will have failed to update it. Could you try force
> pushing (note: this is generally discouraged, but necessary are rebasing a
> branch and pushing to an existing remote branch).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1026584383>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CMABKZX6UDDOMD3D23UY6KHRANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
- Jeff Parker

jeffdparker commented:
I'm looking at my system, and I see no trace of my changes or my pull
request in the exercism branch.

1) I believe that I made the changes, but in a parallel repository, csci1,
with the same problem statement.

2) If Erik thinks there is a problem, what does he think the solution is?
Why would it fail after rebasing?

3) If I didn't make the changes to the exercism branch, do you have any
advice on the next steps?

jeff parker



On Thu, Feb 3, 2022 at 10:11 AM Jeff Parker ***@***.***> wrote:

> Erik
>
> We are out of town: I will take a look at this later this weekend
>
> Jeff
>
> On Tue, Feb 1, 2022 at 1:25 AM Erik Schierboom ***@***.***>
> wrote:
>
>> I think ynfle's suggestion has one tiny bug, and that is that git push
>> will fail after rebasing. Once you've rebased, you'll have to do git
>> push --force. The fact that the contents in this PR are not updated
>> makes sense as the push will have failed to update it. Could you try force
>> pushing (note: this is generally discouraged, but necessary are rebasing a
>> branch and pushing to an existing remote branch).
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1026584383>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEHO6CMABKZX6UDDOMD3D23UY6KHRANCNFSM4HAWDTWA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
> --
> - Jeff Parker
>

ErikSchierboom commented:
> 1) I believe that I made the changes, but in a parallel repository, csci1,
> with the same problem statement.

This is likely the problem. 

> 2) If Erik thinks there is a problem, what does he think the solution is?
> Why would it fail after rebasing?

That's git protecting you from overwriting remote changes with local changes. In essence: `git rebase` will have changed the history which means that git has no way of merging the remote changes with the local changes because they are two different trees.

> 3) If I didn't make the changes to the exercism branch, do you have any
> advice on the next steps?

The branch for this PR is in this repo: https://github.com/jeffdparker/problem-specifications/tree/patch-2 The changes should be applied to the branch of this repo and then pushed.

jeffdparker commented:
Thanks, Erik

On Tue, Feb 8, 2022 at 1:20 AM Erik Schierboom ***@***.***>
wrote:

>
>    1. I believe that I made the changes, but in a parallel repository,
>    csci1,
>    with the same problem statement.
>
> This is likely the problem.
>
>
>    1. If Erik thinks there is a problem, what does he think the solution
>    is?
>    Why would it fail after rebasing?
>
> That's git protecting you from overwriting remote changes with local
> changes. In essence: git rebase will have changed the history which means
> that git has no way of merging the remote changes with the local changes
> because they are two different trees.
>
>
>    1. If I didn't make the changes to the exercism branch, do you have any
>    advice on the next steps?
>
> The branch for this PR is in this repo:
> https://github.com/jeffdparker/problem-specifications/tree/patch-2 The
> changes should be applied to the branch of this repo and then pushed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1032332069>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6COXJS5UCUAC2YZJDJLU2DG5LANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jeffdparker commented:
Erik -

I have edited the file and pushed to patch-2.  Let me know if more is
needed.

I'm curious to know how I could check the syntax before committing.  I did
it by eye, but would prefer some validation

Jeff

On Tue, Feb 8, 2022 at 1:20 AM Erik Schierboom ***@***.***>
wrote:

>
>    1. I believe that I made the changes, but in a parallel repository,
>    csci1,
>    with the same problem statement.
>
> This is likely the problem.
>
>
>    1. If Erik thinks there is a problem, what does he think the solution
>    is?
>    Why would it fail after rebasing?
>
> That's git protecting you from overwriting remote changes with local
> changes. In essence: git rebase will have changed the history which means
> that git has no way of merging the remote changes with the local changes
> because they are two different trees.
>
>
>    1. If I didn't make the changes to the exercism branch, do you have any
>    advice on the next steps?
>
> The branch for this PR is in this repo:
> https://github.com/jeffdparker/problem-specifications/tree/patch-2 The
> changes should be applied to the branch of this repo and then pushed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1032332069>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6COXJS5UCUAC2YZJDJLU2DG5LANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jeffdparker commented:
Erik -

I see that there is an automated check.

There were a couple of commas missing - one in a prior checkin.  I've tried
to fix and recommit

Jeff

On Tue, Feb 8, 2022 at 1:20 AM Erik Schierboom ***@***.***>
wrote:

>
>    1. I believe that I made the changes, but in a parallel repository,
>    csci1,
>    with the same problem statement.
>
> This is likely the problem.
>
>
>    1. If Erik thinks there is a problem, what does he think the solution
>    is?
>    Why would it fail after rebasing?
>
> That's git protecting you from overwriting remote changes with local
> changes. In essence: git rebase will have changed the history which means
> that git has no way of merging the remote changes with the local changes
> because they are two different trees.
>
>
>    1. If I didn't make the changes to the exercism branch, do you have any
>    advice on the next steps?
>
> The branch for this PR is in this repo:
> https://github.com/jeffdparker/problem-specifications/tree/patch-2 The
> changes should be applied to the branch of this repo and then pushed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1032332069>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6COXJS5UCUAC2YZJDJLU2DG5LANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ynfle commented:
Hi @jeffdparker,

It seems that your push didn't fix it.

Check out https://github.com/exercism/problem-specifications#automated-tests for running at least some of the tests locally.
You can also checkout the GitHub Actions CI at https://github.com/exercism/problem-specifications/blob/main/.github/workflows/ci.yml if you're willing to dive in to that

Good luck

ynfle commented:
It seems that `property` should be an array with `[]` around the 2 strings

ErikSchierboom commented on exercises/high-scores/canonical-data.json:
```diff
@@ -112,6 +112,30 @@
           },
           "expected": [25]
         }
```
There's a comma missing here.

ErikSchierboom reviewed (commented):
Just one comma missing I believe.

ErikSchierboom commented on exercises/high-scores/canonical-data.json:
```diff
@@ -103,6 +103,38 @@
+        },
+        {
+          "description": "Recent after a call to sort",
+          "property": "personalTopThree", "latest",
```
This is not valid JSON. Is this test case still needed? I sort of assumed that the two newer ones would handle this.

ErikSchierboom replied:
The `property` value must be a string. This looks like an array, but without the `[` and `]` characters.

jeffdparker commented:
> This is not valid JSON.

Sorry - what is the issue with this?

Is it "scenarios"?

Is it the two properties on the last line?  Was that something I changed?
It does not appear in the most recent diffs

> Is this test case still needed? I sort of assumed that the two newer ones
would handle this.

There are two issues: immutability after calling latest, and immutability
after calling top three.  The old test checked immutability after calling
personal_best(), while the new tests check immutability after calling
personal_top_three().  We should check immutability in both cases: you
could argue that we don't need two tests to check the one thing.

The tests come in pairs: one of these tests checks the most recent (last)
element of the list.  The other test in each pair asks for the full list,
and checks that it is unchanged.  The tests that use the latest score have
the advantage that they don't need access beyond the API.

Jeff


On Thu, Feb 10, 2022 at 12:30 AM Erik Schierboom ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In exercises/high-scores/canonical-data.json
> <https://github.com/exercism/problem-specifications/pull/1486#discussion_r803372868>
> :
>
> > @@ -103,6 +103,38 @@
>            },
>            "expected": [30, 50, 20, 70],
>            "scenarios": ["immutable"]
> +        },
> +        {
> +          "description": "Recent after a call to sort",
> +          "property": "personalTopThree", "latest",
>
> This is not valid JSON. Is this test case still needed? I sort of assumed
> that the two newer ones would handle this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#pullrequestreview-878424808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CMAKIZVFERGUFNCCXDU2NSRRANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ErikSchierboom reviewed (commented):
> The tests come in pairs: one of these tests checks the most recent (last)
> element of the list.  The other test in each pair asks for the full list,
> and checks that it is unchanged.  The tests that use the latest score have
> the advantage that they don't need access beyond the API.

Correct, but the PR is now adding three tests. I think the first one should be removed, right?

jeffdparker commented:
Agreed.

On Fri, Feb 11, 2022 at 12:48 AM Erik Schierboom ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> The tests come in pairs: one of these tests checks the most recent (last)
> element of the list. The other test in each pair asks for the full list,
> and checks that it is unchanged. The tests that use the latest score have
> the advantage that they don't need access beyond the API.
>
> Correct, but the PR is now adding three tests. I think the first one
> should be removed, right?
> ------------------------------
>
> In exercises/high-scores/canonical-data.json
> <https://github.com/exercism/problem-specifications/pull/1486#discussion_r804415683>
> :
>
> > @@ -103,6 +103,38 @@
>            },
>            "expected": [30, 50, 20, 70],
>            "scenarios": ["immutable"]
> +        },
> +        {
> +          "description": "Recent after a call to sort",
> +          "property": "personalTopThree", "latest",
>
> The property value must be a string. This looks like an array, but
> without the [ and ] characters.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#pullrequestreview-879851143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CLCELGQ7KW4IQZHRGLU2S5MZANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jeffdparker commented:
Erik -

I removed the first test, as you suggested, in canonical-data.json

I received an error message about the file, and chided me for not using
prettier.

I have downloaded npm and prettier, and run

npx prettier -c canonical-data.json



on my local file system.  I am told

All matched files use Prettier code style!



Is there a better validation flag I should be using?

jdp

On Fri, Feb 11, 2022 at 12:48 AM Erik Schierboom ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> The tests come in pairs: one of these tests checks the most recent (last)
> element of the list. The other test in each pair asks for the full list,
> and checks that it is unchanged. The tests that use the latest score have
> the advantage that they don't need access beyond the API.
>
> Correct, but the PR is now adding three tests. I think the first one
> should be removed, right?
> ------------------------------
>
> In exercises/high-scores/canonical-data.json
> <https://github.com/exercism/problem-specifications/pull/1486#discussion_r804415683>
> :
>
> > @@ -103,6 +103,38 @@
>            },
>            "expected": [30, 50, 20, 70],
>            "scenarios": ["immutable"]
> +        },
> +        {
> +          "description": "Recent after a call to sort",
> +          "property": "personalTopThree", "latest",
>
> The property value must be a string. This looks like an array, but
> without the [ and ] characters.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#pullrequestreview-879851143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CLCELGQ7KW4IQZHRGLU2S5MZANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jeffdparker commented:
Erik -

I had added the new tests as part of a case statement for the top-three
API.

I have moved the tests two braces out, and the system is silently dreaming
up a new challenge.

jdp

On Fri, Feb 11, 2022 at 12:48 AM Erik Schierboom ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> The tests come in pairs: one of these tests checks the most recent (last)
> element of the list. The other test in each pair asks for the full list,
> and checks that it is unchanged. The tests that use the latest score have
> the advantage that they don't need access beyond the API.
>
> Correct, but the PR is now adding three tests. I think the first one
> should be removed, right?
> ------------------------------
>
> In exercises/high-scores/canonical-data.json
> <https://github.com/exercism/problem-specifications/pull/1486#discussion_r804415683>
> :
>
> > @@ -103,6 +103,38 @@
>            },
>            "expected": [30, 50, 20, 70],
>            "scenarios": ["immutable"]
> +        },
> +        {
> +          "description": "Recent after a call to sort",
> +          "property": "personalTopThree", "latest",
>
> The property value must be a string. This looks like an array, but
> without the [ and ] characters.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#pullrequestreview-879851143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CLCELGQ7KW4IQZHRGLU2S5MZANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ErikSchierboom commented:
/format

loziniak commented:
seems already addressed by #1767 .

jeffdparker commented:
> seems already addressed by #1767
<https://github.com/exercism/problem-specifications/pull/1767> .

There are two calls in the API that need to be checked: this change checks
the other one.  As noted above:

> Briefly, there are three calls: latest(), ..._best() and
 ..._top_three().  There is no temptation to modify the list in
implementing latest(), but there is for the other two, and we should check
that calling each does not modify the list...

On Tue, Feb 15, 2022 at 3:17 AM Erik Schierboom ***@***.***>
wrote:

> seems already addressed by #1767
> <https://github.com/exercism/problem-specifications/pull/1767> .
>
> This is a new addition.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/problem-specifications/pull/1486#issuecomment-1040094678>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHO6CL5X6Z372WF3UYN56TU3IR23ANCNFSM4HAWDTWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
